### PR TITLE
Added ArcTan2 operation

### PR DESCRIPTION
**Pull request by yrahul3910**
## Proposed changes

This PR adds the arctan2 operation, as discussed in #1065.

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

### CODE REVIEW

File: mlx/backend/accelerate/primitives.cpp [hunk 1]
```diff
@@ -193,6 +193,25 @@ void ArcTan::eval_cpu(const std::vector<array>& inputs, array& out) {
...
+    } else if (b.is_donatable() && b.itemsize() == out.itemsize()) {
+      out.copy_shared_buffer(b);
+    } else {
+      out.set_data(allocator::malloc_or_wait(out.nbytes()));
+    }
+    int size = a.data_size();
+    vvatan2f(out.data<float>(), a.data<float>(), b.data<float>(), &size);
+  } else {
+    eval(inputs, out);
+  }
```
Comment (awni) [inline mlx/backend/accelerate/primitives.cpp:212]: This is not correct. Could you follow the example of `Add` (or any other binary primitive)?

File: mlx/backend/common/primitives.cpp [hunk 2]
```diff
@@ -94,6 +94,19 @@ void ArcTan::eval(const std::vector<array>& inputs, array& out) {
...
+  const auto& in1 = inputs[0];
+  const auto& in2 = inputs[1];
+  if (issubdtype(out.dtype(), inexact)) {
+    binary(in1, in2, out, detail::ArcTan2());
+  } else {
+    throw std::invalid_argument(
+        "[arctan2] Cannot compute inverse tangent of elements in array"
+        " with non floating point type.");
+  }
+}
```
Comment (awni) [inline mlx/backend/common/primitives.cpp:108]: Move this to `binary.cpp` and follow the example of `LogAddExp` to only compile for floating point types: https://github.com/ml-explore/mlx/blob/main/mlx/backend/common/binary.cpp#L178-L197

File: mlx/primitives.cpp [hunk 3]
```diff
@@ -5,6 +5,7 @@
 #include <numeric>
 #include <sstream>
 #include <stdexcept>
+#include <utility>
```
Comment (awni) [inline mlx/primitives.cpp:8]: Was this added intentionally?

File: mlx/primitives.h [hunk 4]
```diff
@@ -314,6 +314,23 @@ class ArcTan : public UnaryPrimitive {
   void eval(const std::vector<array>& inputs, array& out);
 };
 
+class ArcTan2 : public UnaryPrimitive {
```
Comment (awni) [inline mlx/primitives.h:317]: Could you add it to this list as well: https://github.com/ml-explore/mlx/blob/main/mlx/compile.cpp#L38

File: mlx/ops.h [hunk 5]
```diff
@@ -828,6 +828,9 @@ array arccos(const array& a, StreamOrDevice s = {});
 /** Arc Tangent of the elements of an array */
 array arctan(const array& a, StreamOrDevice s = {});
 
+/** Arc Tangent of elements of two arrays */
```
Comment (awni) [inline mlx/ops.h:831]: Clarify this docstring please: `Inverse tangent of the ratio of two arrays`. Also please use `a` and `b` here to follow the Python API.

**Review by awni — CHANGES_REQUESTED** [review]
Thanks for adding this. I left a few comments inline

Also you'll need to add the primitive to this list to enable compilation in CPU only mode: https://github.com/ml-explore/mlx/blob/main/mlx/backend/no_metal/primitives.cpp

File: mlx/ops.cpp [hunk 6]
```diff
@@ -3,6 +3,7 @@
 #include <algorithm>
 #include <climits>
 #include <cmath>
+#include <cwchar>
```
[the remark this answers is not in the document]
Comment (yrahul3910) [reply mlx/ops.cpp:6]: Whoops, seems there were a couple of these. I'll get rid of them.

File: mlx/backend/accelerate/primitives.cpp (hunk 1 above)
Comment (yrahul3910) [reply mlx/backend/accelerate/primitives.cpp:212]: Ah, gotcha. I was following `Power`. I'll update it now :)

Comment (awni) [reply mlx/backend/accelerate/primitives.cpp:212]: Actually following `Power` might be better given there are no scalar equivalents. However, you would need to use `row_contiguous` here and not `contiguous` (which is a bug).

Comment (yrahul3910) [reply mlx/backend/accelerate/primitives.cpp:212]: Ah, I was stuck on that and just about to ask you :) I'll re-run tests/lint and push the commit to the PR.

File: mlx/backend/accelerate/primitives.cpp [hunk 7]
```diff
@@ -193,6 +193,26 @@ void ArcTan::eval_cpu(const std::vector<array>& inputs, array& out) {
...
 
+void ArcTan2::eval_cpu(const std::vector<array>& inputs, array& out) {
+  assert(inputs.size() == 2);
+  auto& a = inputs[0];
+  auto& b = inputs[1];
+  if (out.dtype() == float32 && a.flags().row_contiguous &&
+      b.flags().row_contiguous) {
+    if (a.is_donatable() && a.itemsize() == out.itemsize()) {
+      out.copy_shared_buffer(a);
+    } else if (b.is_donatable() && b.itemsize() == out.itemsize()) {
```
Comment (awni) [inline mlx/backend/accelerate/primitives.cpp:204]: I don't think we need to check the `itemsize` here. `a` and `b` have to have the same type as `out` at this point.

File: mlx/primitives.cpp [hunk 8]
```diff
@@ -402,6 +402,37 @@ std::pair<std::vector<array>, std::vector<int>> ArcTan::vmap(
...
+  return jvp(primals, cotangents, argnums);
+}
+
+std::vector<array> ArcTan2::jvp(
+    const std::vector<array>& primals,
+    const std::vector<array>& tangents,
+    const std::vector<int>& argnums) {
+  assert(primals.size() == 2);
+  assert(argnums.size() == 2);
+  array one = array(1., primals[0].dtype());
```
Comment (awni) [inline mlx/primitives.cpp:419]: Unused?

File: mlx/backend/common/binary.cpp [hunk 9]
```diff
@@ -293,4 +293,25 @@ void BitwiseBinary::eval_cpu(const std::vector<array>& inputs, array& out) {
...
+  if (out.dtype() == float32) {
+    binary_op<float>(a, b, out, detail::ArcTan2());
+  } else if (out.dtype() == float16) {
+    binary_op<float16_t>(a, b, out, detail::ArcTan2());
+  } else if (out.dtype() == bfloat16) {
+    binary_op<bfloat16_t>(a, b, out, detail::ArcTan2());
+  } else if (issubdtype(out.dtype(), inexact)) {
+    std::ostringstream err;
+    err << "[arctan2] Does not support " << out.dtype();
+    throw std::invalid_argument(err.str());
```
Comment (awni) [inline mlx/backend/common/binary.cpp:309]: Did you try for `complex`? Does `std::atan2` not support it?

**Review by awni — COMMENTED** [review]
This looks great! I left just a couple more comments then should be good to go.

File: mlx/primitives.cpp (hunk 8 above)
Comment (yrahul3910) [reply mlx/primitives.cpp:419]: I wasn't very sure--I added it since `jvp` and `vjp` are `virtual` in `Primitive`

Comment (awni) [reply mlx/primitives.cpp:419]: I meant the `array one` not the function itself which is should be added.

Comment (yrahul3910) [reply mlx/primitives.cpp:419]: Ooh, yes, nice catch, thanks!

File: mlx/backend/common/binary.cpp (hunk 9 above)
Comment (yrahul3910) [reply mlx/backend/common/binary.cpp:309]: I don't think so: the [docs for atan2](https://en.cppreference.com/w/cpp/numeric/math/atan2) don't seem to have an overload for it, and the [complex class has their own inverse trig functions, but not atan2](https://en.cppreference.com/w/cpp/numeric/complex/atan). Perhaps we could roll our own?

**Comment by yrahul3910** [comment]
Thanks! Is that linux build supposed to pass? It seems like it's mainly having trouble finding the ArcTan2 symbol and I'm not sure what I missed.

**Comment by awni** [comment]
Yes, linux tests should pass. I think you just need to add  to the list here: https://github.com/ml-explore/mlx/blob/main/mlx/backend/common/default_primitives.cpp#L36

That is just a way of saying that we should just use the default `eval` when accelerate is not available (e.g. on linux).

**Review by awni — APPROVED** [review]
Looks great, thanks!!

**Pull request merged**